src/core/ext/filters/deadline/deadline_filter.cc

```diff
@@ -128,21 +128,25 @@ static void cancel_timer_if_needed(grpc_deadline_state* deadline_state) {
   }
 }
 
-// Callback run when the call is complete.
-static void on_complete(void* arg, grpc_error* error) {
+// Callback run when we receive trailing metadata.
+static void recv_trailing_metadata_ready(void* arg, grpc_error* error) {
   grpc_deadline_state* deadline_state = static_cast<grpc_deadline_state*>(arg);
   cancel_timer_if_needed(deadline_state);
-  // Invoke the next callback.
-  GRPC_CLOSURE_RUN(deadline_state->next_on_complete, GRPC_ERROR_REF(error));
+  // Invoke the original callback.
+  GRPC_CLOSURE_RUN(deadline_state->original_recv_trailing_metadata_ready,
+                   GRPC_ERROR_REF(error));
 }
 
-// Inject our own on_complete callback into op.
-static void inject_on_complete_cb(grpc_deadline_state* deadline_state,
-                                  grpc_transport_stream_op_batch* op) {
-  deadline_state->next_on_complete = op->on_complete;
-  GRPC_CLOSURE_INIT(&deadline_state->on_complete, on_complete, deadline_state,
+// Inject our own recv_trailing_metadata_ready callback into op.
+static void inject_recv_trailing_metadata_ready(
+    grpc_deadline_state* deadline_state, grpc_transport_stream_op_batch* op) {
+  deadline_state->original_recv_trailing_metadata_ready =
+      op->payload->recv_trailing_metadata.recv_trailing_metadata_ready;
```

Comment:
The recv_* closures are op specific, whereas the on_complete closure **used to be** batch specific. You make a very good point here. Now on_complete is only send op specific, so it could live in payload, though I am not sure how the details of that would work

Reply:
When a call has multiple batches in flight at the same time, all of the batches share the same payload.  There can be only one of each recv op in flight at a given time, so there's no conflict if we set the recv callbacks in the payload, because only one of the pending batches can contain a given recv op (and therefore need to set the recv op's callback).  However, there can be multiple batches containing send ops, and each one will need a different on_complete closure.  Therefore, the on_complete closure cannot be in the payload.

Reply:
So to answer the question, putting the recv callbacks in the payload is a memory allocation optimization, because all the batches share one allocated payload.